File: android-app/src/main/java/com/prevengos/plug/android/ui/CaptureViewModelFactory.java
package com.prevengos.plug.android.ui;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.prevengos.plug.android.PrevengosApplication;
import com.prevengos.plug.android.di.AppContainer;

public class CaptureViewModelFactory implements ViewModelProvider.Factory {
    private final PrevengosApplication application;

    public CaptureViewModelFactory(PrevengosApplication application) {
        this.application = application;
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(CaptureViewModel.class)) {
            AppContainer container = application.getContainer();
            return (T) new CaptureViewModel(
                    container.getPacienteRepository(),
                    container.getCuestionarioRepository(),
                    application,
                    container.getIoExecutor());
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: android-app/src/main/java/com/prevengos/plug/android/ui/CaptureViewModelFactory.kt
package com.prevengos.plug.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.prevengos.plug.android.PrevengosApplication

class CaptureViewModelFactory(private val application: PrevengosApplication) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(CaptureViewModel::class.java)) {
            val container = application.container
            return CaptureViewModel(
                pacienteRepository = container.pacienteRepository,
                cuestionarioRepository = container.cuestionarioRepository,
                application = application
            ) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
